pkg/services/object/get/ec.go

```diff
@@ -0,0 +1,399 @@
+				)
+				return nil
+			}
+
+			parts[partIdx] = part
+			if okCounter.Add(1) >= uint32(rem) {
+				return errInterrupt
+			}
+
+			return nil
+		})
+	}
+	if err := eg.Wait(); err != nil && !errors.Is(err, errInterrupt) {
+		return object.Object{}, err
+	}
+
+	if rem = islices.CountNilsInTwoDimSlice(parts); rem > int(rule.ParityPartNum) {
+		return object.Object{}, tooManyPartsUnavailableError(rem)
+	}
+
+	payload, err := iec.Decode(rule, hdr.PayloadSize(), parts)
+	if err != nil {
+		return object.Object{}, fmt.Errorf("decode payload from parts: %w", err)
+	}
+
+	hdr.SetPayload(payload)
+	return hdr, nil
+}
+
+// reads object for EC part which should produced for parent object and
+// distributed sortedNodes according to EC rule with index = ruleIdx in cnr
+// policy. Returns parent object header and part payload. Payload slice is nil
+// iff parent payload is empty.
+//
+// Returns [apistatus.ErrObjectAlreadyRemoved] if the object was marked for
+// removal.
+//
+// Can return [context.Canceled] from the passed ctx only.
+func (s *Service) getECPart(ctx context.Context, cnr cid.ID, parent oid.ID, sTok *session.Object, bTok *bearer.Token,
+	rule iec.Rule, ruleIdx int, sortedNodes []netmap.NodeInfo, partIdx int) (object.Object, []byte, error) {
+	var partHdr object.Object
+	var rc io.ReadCloser
+
+	pi := iec.PartInfo{
+		RuleIndex: ruleIdx,
+		Index:     partIdx,
+	}
+
+	var err error
+	var local, localDone bool
+	for i := range iec.NodeSequenceForPart(partIdx, int(rule.DataPartNum+rule.ParityPartNum), len(sortedNodes)) {
+		local = !localDone && s.neoFSNet.IsLocalNodePublicKey(sortedNodes[i].PublicKey())
+		if local {
+			localDone = true
+		}
+
+		if local {
+			partHdr, rc, err = s.localObjects.GetECPart(cnr, parent, pi)
+		} else {
+			partHdr, rc, err = s.getECPartFromNode(ctx, cnr, parent, sTok, bTok, pi, sortedNodes[i])
+		}
+		if err == nil {
+			break
+		}
+		if errors.Is(err, apistatus.ErrObjectAlreadyRemoved) {
+			return object.Object{}, nil, err
+		}
+		if errors.Is(err, ctx.Err()) {
+			return object.Object{}, nil, err
+		}
+
+		if !errors.Is(err, apistatus.ErrObjectNotFound) {
+			s.log.Info("failed to get EC part from node, continue...",
+				zap.Stringer("container", cnr), zap.Stringer("object", parent), zap.Stringer("rule", rule),
+				zap.Int("ruleIdx", pi.RuleIndex), zap.Int("partIdx", pi.Index), zap.Bool("local", local), zap.Error(err))
+		}
+	}
+	if err != nil {
+		return object.Object{}, nil, errors.New("all nodes failed")
+	}
+
+	defer rc.Close()
+
+	parentHdr := partHdr.Parent()
+	if parentHdr == nil {
+		return object.Object{}, nil, errors.New("missing parent header in object for part")
+	}
+
+	if partHdr.PayloadSize() > parentHdr.PayloadSize() {
+		return object.Object{}, nil, errors.New("part object payload is bigger than the parent one")
+	}
+
+	if parentHdr.PayloadSize() == 0 {
+		return *parentHdr, nil, nil
+	}
+
+	buf := make([]byte, partHdr.PayloadSize())
```

Comment:
can it be gigabytes?

Reply:
it can be up to `MaxObjectSize` net setting

Reply:
oh, missed that there are `part` and `parent` headers (confusing naming btw, at least, i misunderstood)

> it can be up to MaxObjectSize net setting

ok, but even in this case, is it ok to always allocate so much memory? found TODOs about streaming but isn't it necessary to solve ASAP?

Reply:
> isn't it necessary to solve ASAP

no. And not only due to not yet resolved streaming issue, but cuz decoder can now work in-mem only (#3501)